File: taurus-protect-sdk-go/pkg/protect/crypto/tpv1_test.go
package crypto

import (
	"net/http"
	"strings"
	"testing"
)

func TestNewTPV1Auth(t *testing.T) {
	tests := []struct {
		name      string
		apiKey    string
		apiSecret string
		wantErr   bool
	}{
		{
			name:      "valid credentials",
			apiKey:    "test-api-key",
			apiSecret: "deadbeef",
			wantErr:   false,
		},
		{
			name:      "empty api key",
			apiKey:    "",
			apiSecret: "deadbeef",
			wantErr:   true,
		},
		{
			name:      "empty api secret",
			apiKey:    "test-api-key",
			apiSecret: "",
			wantErr:   true,
		},
		{
			name:      "invalid hex secret",
			apiKey:    "test-api-key",
			apiSecret: "not-hex",
			wantErr:   true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			auth, err := NewTPV1Auth(tt.apiKey, tt.apiSecret)
			if (err != nil) != tt.wantErr {
				t.Errorf("NewTPV1Auth() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !tt.wantErr && auth == nil {
				t.Error("NewTPV1Auth() returned nil auth without error")
			}
			if auth != nil {
				auth.Close()
			}
		})
	}
}

func TestTPV1Auth_SignRequest(t *testing.T) {
	auth, err := NewTPV1Auth("test-api-key", "deadbeefcafe1234")
	if err != nil {
		t.Fatalf("Failed to create TPV1Auth: %v", err)
	}
	defer auth.Close()

	req, err := http.NewRequest("GET", "https://api.example.com/api/rest/v1/wallets", nil)
	if err != nil {
		t.Fatalf("Failed to create request: %v", err)
	}

	err = auth.SignRequest(req, nil)
	if err != nil {
		t.Errorf("SignRequest() error = %v", err)
	}

	authHeader := req.Header.Get("Authorization")
	if authHeader == "" {
		t.Error("Authorization header not set")
	}

	if !strings.HasPrefix(authHeader, "TPV1-HMAC-SHA256 ") {
		t.Errorf("Authorization header has wrong prefix: %s", authHeader)
	}

	if !strings.Contains(authHeader, "ApiKey=test-api-key") {
		t.Errorf("Authorization header missing ApiKey: %s", authHeader)
	}

	if !strings.Contains(authHeader, "Nonce=") {
		t.Errorf("Authorization header missing Nonce: %s", authHeader)
	}

	if !strings.Contains(authHeader, "Timestamp=") {
		t.Errorf("Authorization header missing Timestamp: %s", authHeader)
	}

	if !strings.Contains(authHeader, "Signature=") {
		t.Errorf("Authorization header missing Signature: %s", authHeader)
	}
}

func TestTPV1Auth_SignRequestWithBody(t *testing.T) {
	auth, err := NewTPV1Auth("test-api-key", "deadbeefcafe1234")
	if err != nil {
		t.Fatalf("Failed to create TPV1Auth: %v", err)
	}
	defer auth.Close()

	body := []byte(`{"name":"test-wallet"}`)
	req, err := http.NewRequest("POST", "https://api.example.com/api/rest/v1/wallets", nil)
	if err != nil {
		t.Fatalf("Failed to create request: %v", err)
	}
	req.Header.Set("Content-Type", "application/json")

	err = auth.SignRequest(req, body)
	if err != nil {
		t.Errorf("SignRequest() error = %v", err)
	}

	// Content-Type should be normalized
	contentType := req.Header.Get("Content-Type")
	if contentType != "application/json; charset=utf-8" {
		t.Errorf("Content-Type not normalized: %s", contentType)
	}

	authHeader := req.Header.Get("Authorization")
	if authHeader == "" {
		t.Error("Authorization header not set")
	}
}

func TestCalculateSignedHeader(t *testing.T) {
	apiSecret := []byte{0xde, 0xad, 0xbe, 0xef}
	header := CalculateSignedHeader(
		"test-key",
		apiSecret,
		"test-nonce",
		1234567890,
		"GET",
		"api.example.com",
		"/api/v1/test",
		"",
		"",
		"",
	)

	if !strings.HasPrefix(header, "TPV1-HMAC-SHA256 ") {
		t.Errorf("Header has wrong prefix: %s", header)
	}

	if !strings.Contains(header, "ApiKey=test-key") {
		t.Errorf("Header missing ApiKey: %s", header)
	}

	if !strings.Contains(header, "Nonce=test-nonce") {
		t.Errorf("Header missing Nonce: %s", header)
	}

	if !strings.Contains(header, "Timestamp=1234567890") {
		t.Errorf("Header missing Timestamp: %s", header)
	}
}

func TestCalculateBase64HMAC(t *testing.T) {
	secret := []byte("test-secret")
	data := "test-data"

	hmac1 := CalculateBase64HMAC(secret, data)
	hmac2 := CalculateBase64HMAC(secret, data)

	if hmac1 != hmac2 {
		t.Error("HMAC should be deterministic")
	}

	if !CheckBase64HMAC(secret, data, hmac1) {
		t.Error("CheckBase64HMAC should return true for valid HMAC")
	}

	if CheckBase64HMAC(secret, data, "invalid") {
		t.Error("CheckBase64HMAC should return false for invalid HMAC")
	}
}

func TestCalculateHexHash(t *testing.T) {
	hash := CalculateHexHash("test")
	// SHA256 of "test"
	expected := "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"
	if hash != expected {
		t.Errorf("CalculateHexHash() = %s, want %s", hash, expected)
	}
}

func TestWipe(t *testing.T) {
	data := []byte{1, 2, 3, 4, 5}
	Wipe(data)

	for i, b := range data {
		if b != 0 {
			t.Errorf("Wipe() did not zero byte at index %d: got %d", i, b)
		}
	}
}

func TestWipe_EmptySlice(t *testing.T) {
	var data []byte
	// Should not panic
	Wipe(data)
}

func TestTPV1Auth_Close_WipesSecret(t *testing.T) {
	auth, err := NewTPV1Auth("test-api-key", "deadbeef0123456789abcdef01234567")
	if err != nil {
		t.Fatalf("NewTPV1Auth() error = %v", err)
	}

	// Capture secret length before close
	secretLen := len(auth.apiSecret)
	if secretLen == 0 {
		t.Fatal("apiSecret should not be empty")
	}

	auth.Close()

	// Verify secret is wiped (all zeros)
	for i, b := range auth.apiSecret {
		if b != 0 {
			t.Errorf("apiSecret[%d] = %v after Close(), want 0", i, b)
		}
	}
}

func TestDecodePublicKeyPEM_ErrorCases(t *testing.T) {
	tests := []struct {
		name    string
		pem     string
		wantErr bool
	}{
		{"empty PEM", "", true},
		{"invalid PEM", "not a pem", true},
		{"invalid PEM header", "-----BEGIN SOMETHING-----\naW52YWxpZA==\n-----END SOMETHING-----", true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := DecodePublicKeyPEM(tt.pem)
			if (err != nil) != tt.wantErr {
				t.Errorf("DecodePublicKeyPEM() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestDecodePrivateKeyPEM_ErrorCases(t *testing.T) {
	tests := []struct {
		name    string
		pem     string
		wantErr bool
	}{
		{"empty PEM", "", true},
		{"invalid PEM", "not a pem", true},
		{"invalid PEM content", "-----BEGIN EC PRIVATE KEY-----\naW52YWxpZA==\n-----END EC PRIVATE KEY-----", true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := DecodePrivateKeyPEM(tt.pem)
			if (err != nil) != tt.wantErr {
				t.Errorf("DecodePrivateKeyPEM() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestNewTPV1Auth_ErrorMessages(t *testing.T) {
	tests := []struct {
		name         string
		apiKey       string
		apiSecretHex string
		errContains  string
	}{
		{
			name:         "empty api key",
			apiKey:       "",
			apiSecretHex: "deadbeef",
			errContains:  "apiKey cannot be empty",
		},
		{
			name:         "empty api secret",
			apiKey:       "test-key",
			apiSecretHex: "",
			errContains:  "apiSecret cannot be empty",
		},
		{
			name:         "invalid hex",
			apiKey:       "test-key",
			apiSecretHex: "not-valid-hex!",
			errContains:  "invalid apiSecret",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := NewTPV1Auth(tt.apiKey, tt.apiSecretHex)
			if err == nil {
				t.Error("Expected error, got nil")
				return
			}
			if !strings.Contains(err.Error(), tt.errContains) {
				t.Errorf("Error = %v, want containing %q", err, tt.errContains)
			}
		})
	}
}

File: taurus-protect-sdk-typescript/tests/unit/crypto/tpv1.test.ts
/**
 * Unit tests for TPV1-HMAC-SHA256 authentication.
 *
 * These tests verify the TPV1Auth class and calculateSignedHeader function
 * used for API request authentication.
 */

import {
  calculateHexHash,
  calculateBase64Hmac,
  verifyBase64Hmac,
} from "../../../src/crypto/hashing";
import { TPV1Auth, calculateSignedHeader } from "../../../src/crypto/tpv1";

describe("calculateHexHash (TPV1 context)", () => {
  it("should produce expected SHA-256 for known input", () => {
    const hash = calculateHexHash("hello");
    expect(hash).toBe(
      "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824"
    );
  });

  it("should produce expected SHA-256 for empty string", () => {
    const hash = calculateHexHash("");
    expect(hash).toBe(
      "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
    );
  });
});

describe("calculateBase64Hmac", () => {
  it("should produce a deterministic HMAC for known inputs", () => {
    const secret = Buffer.from("secret-key");
    const hmac1 = calculateBase64Hmac(secret, "data");
    const hmac2 = calculateBase64Hmac(secret, "data");
    expect(hmac1).toBe(hmac2);
  });

  it("should produce different HMACs for different secrets", () => {
    const hmac1 = calculateBase64Hmac(Buffer.from("key-a"), "data");
    const hmac2 = calculateBase64Hmac(Buffer.from("key-b"), "data");
    expect(hmac1).not.toBe(hmac2);
  });
});

describe("verifyBase64Hmac", () => {
  it("should return true for matching HMAC", () => {
    const secret = Buffer.from("my-secret");
    const data = "some-data-to-sign";
    const hmac = calculateBase64Hmac(secret, data);
    expect(verifyBase64Hmac(secret, data, hmac)).toBe(true);
  });

  it("should return false for non-matching HMAC", () => {
    const secret = Buffer.from("my-secret");
    expect(verifyBase64Hmac(secret, "data", "wrong-hmac-value")).toBe(false);
  });

  it("should return false when data is different", () => {
    const secret = Buffer.from("my-secret");
    const hmac = calculateBase64Hmac(secret, "original-data");
    expect(verifyBase64Hmac(secret, "tampered-data", hmac)).toBe(false);
  });
});

describe("TPV1Auth", () => {
  const apiKey = "test-api-key";
  const apiSecretHex = "deadbeef01020304";

  it("should throw on empty apiKey", () => {
    expect(() => new TPV1Auth("", apiSecretHex)).toThrow(
      "apiKey cannot be empty"
    );
  });

  it("should throw on empty apiSecret", () => {
    expect(() => new TPV1Auth(apiKey, "")).toThrow("apiSecret cannot be empty");
  });

  describe("signRequest", () => {
    let auth: TPV1Auth;

    beforeEach(() => {
      auth = new TPV1Auth(apiKey, apiSecretHex);
    });

    afterEach(() => {
      auth.close();
    });

    it('should produce header starting with "TPV1-HMAC-SHA256"', () => {
      const header = auth.signRequest(
        "GET",
        "api.example.com",
        "/v1/wallets"
      );
      expect(header.startsWith("TPV1-HMAC-SHA256 ")).toBe(true);
    });

    it("should include ApiKey, Nonce, Timestamp, Signature", () => {
      const header = auth.signRequest(
        "GET",
        "api.example.com",
        "/v1/wallets"
      );
      expect(header).toContain(`ApiKey=${apiKey}`);
      expect(header).toMatch(/Nonce=[0-9a-f-]{36}/);
      expect(header).toMatch(/Timestamp=\d+/);
      expect(header).toMatch(/Signature=.+/);
    });

    it("should produce different signatures for GET vs POST", () => {
      const headerGet = auth.signRequest(
        "GET",
        "api.example.com",
        "/v1/wallets"
      );
      const headerPost = auth.signRequest(
        "POST",
        "api.example.com",
        "/v1/wallets"
      );

      // Extract just the signature portion
      const sigGet = headerGet.split("Signature=")[1];
      const sigPost = headerPost.split("Signature=")[1];
      expect(sigGet).not.toBe(sigPost);
    });

    it("should handle GET without body", () => {
      const header = auth.signRequest(
        "GET",
        "api.example.com",
        "/v1/wallets"
      );
      expect(header).toBeDefined();
      expect(header.startsWith("TPV1-HMAC-SHA256 ")).toBe(true);
    });

    it("should handle POST with JSON body", () => {
      const header = auth.signRequest(
        "POST",
        "api.example.com",
        "/v1/wallets",
        undefined,
        "application/json",
        '{"name":"my-wallet"}'
      );
      expect(header).toBeDefined();
      expect(header.startsWith("TPV1-HMAC-SHA256 ")).toBe(true);
    });

    it("should include query string when provided", () => {
      // Sign with and without query to confirm they differ
      const headerNoQuery = auth.signRequest(
        "GET",
        "api.example.com",
        "/v1/wallets"
      );
      const headerWithQuery = auth.signRequest(
        "GET",
        "api.example.com",
        "/v1/wallets",
        "limit=10&offset=0"
      );

      const sigNoQuery = headerNoQuery.split("Signature=")[1];
      const sigWithQuery = headerWithQuery.split("Signature=")[1];
      expect(sigNoQuery).not.toBe(sigWithQuery);
    });

    it("should uppercase the HTTP method", () => {
      // Signing with "get" should produce the same signature as "GET"
      const auth2 = new TPV1Auth(apiKey, apiSecretHex);
      // We can't easily verify method uppercasing since nonce/timestamp differ,
      // but we verify it doesn't throw
      const header = auth.signRequest(
        "get",
        "api.example.com",
        "/v1/wallets"
      );
      expect(header).toBeDefined();
      auth2.close();
    });
  });

  describe("close", () => {
    it("should throw after close when signing", () => {
      const auth = new TPV1Auth(apiKey, apiSecretHex);
      auth.close();
      expect(() =>
        auth.signRequest("GET", "api.example.com", "/v1/wallets")
      ).toThrow("TPV1Auth has been closed");
    });

    it("should be idempotent (calling close twice does not throw)", () => {
      const auth = new TPV1Auth(apiKey, apiSecretHex);
      auth.close();
      expect(() => auth.close()).not.toThrow();
    });
  });

  describe("parseUrl", () => {
    it("should parse a simple URL", () => {
      const result = TPV1Auth.parseUrl("https://api.example.com/v1/wallets");
      expect(result.host).toBe("api.example.com");
      expect(result.path).toBe("/v1/wallets");
      expect(result.query).toBeUndefined();
    });

    it("should parse URL with query string", () => {
      const result = TPV1Auth.parseUrl(
        "https://api.example.com/v1/wallets?limit=10&offset=0"
      );
      expect(result.host).toBe("api.example.com");
      expect(result.path).toBe("/v1/wallets");
      expect(result.query).toBe("limit=10&offset=0");
    });

    it("should handle URL with port", () => {
      const result = TPV1Auth.parseUrl("https://api.example.com:8443/v1/health");
      expect(result.host).toBe("api.example.com:8443");
      expect(result.path).toBe("/v1/health");
    });
  });
});

describe("calculateSignedHeader", () => {
  it("should produce a valid TPV1 header", () => {
    const secret = Buffer.from("deadbeef", "hex");
    const header = calculateSignedHeader(
      "my-key",
      secret,
      "nonce-123",
      1700000000000,
      "GET",
      "api.example.com",
      "/v1/wallets"
    );

    expect(header.startsWith("TPV1-HMAC-SHA256 ")).toBe(true);
    expect(header).toContain("ApiKey=my-key");
    expect(header).toContain("Nonce=nonce-123");
    expect(header).toContain("Timestamp=1700000000000");
    expect(header).toMatch(/Signature=.+/);
  });

  it("should be deterministic for same inputs", () => {
    const secret = Buffer.from("deadbeef", "hex");
    const args = [
      "my-key",
      secret,
      "fixed-nonce",
      1700000000000,
      "POST",
      "api.example.com",
      "/v1/wallets",
      undefined,
      "application/json",
      '{"name":"test"}',
    ] as const;

    const header1 = calculateSignedHeader(...args);
    const header2 = calculateSignedHeader(...args);
    expect(header1).toBe(header2);
  });

  it("should include body in signature computation", () => {
    const secret = Buffer.from("deadbeef", "hex");
    const common = [
      "my-key",
      secret,
      "nonce-1",
      1700000000000,
      "POST",
      "api.example.com",
      "/v1/wallets",
      undefined,
      "application/json",
    ] as const;

    const header1 = calculateSignedHeader(...common, '{"a":1}');
    const header2 = calculateSignedHeader(...common, '{"b":2}');

    const sig1 = header1.split("Signature=")[1];
    const sig2 = header2.split("Signature=")[1];
    expect(sig1).not.toBe(sig2);
  });
});
